### PR TITLE
Gguyo function verification

### PR DESCRIPTION
**Pull request by GemechisUrgessa**
### Description

This PR addresses the following tasks:

1. Debugging the function verification process to ensure accurate results.
2. Adding additional test cases to enhance test coverage and validate the verification of the functions.

### Changes Made

- Debugged the function verification process to resolve any existing issues or inaccuracies.
- Added new test cases to cover some scenarios and validate the functionality of the functions.

### CODE REVIEW

File: compiler/verif/z3refinement.ml [hunk 1]
```diff
@@ -1721,22 +1724,34 @@ and vc_gen_typ_exp_desc ctx env typenv t =
...
-       (* add_constraint env expr; *)
-       z3_solve ctx env expr;
-       )
-  | Erefinementpairfuntype(txp_list, exp) -> debug(Printf.sprintf "Erefinementfunpair \n")
-       (* List.iter (fun elem ->         ) txp_list *)
+  | Erefinement(t, e) -> 
+    debug(Printf.sprintf "Erefinement\n");
+    let expr = vc_gen_expression ctx env e typenv in
+    (* get the variable name from the constriant and replace them by the reference variable *)
+    let expr2 =
```
Comment (jlvargasme) [inline compiler/verif/z3refinement.ml]: I suggest having a helper function to get expr2 instead of writing nested pattern matches. This will help make the codebase easy to read.
There are many other nested pattern matches in the codebase, but that was mostly due to bad coding practices when the project started.

File: test/marvelus/function_test.zls [hunk 2]
```diff
@@ -0,0 +1,87 @@
+(* This is a test file for the function refinement type checker  *)
```
Comment (jlvargasme) [inline test/marvelus/function_test.zls:1]: It seems all function bodies have only one line. Is this a current limitation of the code? If so, what are the issues?

If not, can you add some unit tests with a function body with more lines? Maybe mixing operations and creating some intermediate variables.

File: test/marvelus/function_test.zls [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+(*
+   The multiply function takes two float values, `x` and `y`,
+   and returns a float value `z` such that `z >= x` and `z >= y`.
+   This function is expected to fail the refinement verification.
+    # (=>(and (= r z) (= z (/*/ x y)) (> y 0.0) (> x 0.0)) (or(>= r x) (<=r x) (>= r y) (<=r x))
+   Constraints:
+   - `x` and `y` are integers.
+   - The return value `z` should satisfy the constraints `z >= x` and `z >= y`.
+*)
+let multiply (x : {w:float|w>0} ) (y : {v:float|v>0}) : {r:float | r >= x && r >= y } =
```
Comment (jlvargasme) [inline test/marvelus/function_test.zls:43]: If we mix multiple operations, i.e (z = x + y ** y - y) can the type checker prove the refinement type?

File: test/marvelus/function_test.zls [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  let z = x /. y in
+  z
+
+(* Entry Point *)
+(*
+   The node function serves as the entry point for the test.
+   It calls the `add` , `sub` and `multiply` functions with sample arguments,
+   prints the results, and returns unit ().
+*)
+let node main () =
```
Comment (jlvargasme) [inline test/marvelus/function_test.zls:115]: In a few of the outputs, I can see a "42" magic number (see below). Where are those coming from? 
This usually complicates more complicated proofs and would be good to fix whenever we see them.

```
vc_gen_expression 42
Operator call: print_float
vc_gen_expression 2.0
Operator call: add
map2 error is here!!
name of function:add
length of ref_fun.argument_list:2
# of Arguments in arg_list:2 arguments:2
Proving constraint: (not (=> (and (= 2.0 x) (= 3.0 y)) (and (> y 0.0) (> x 0.0))))
Passed
```

**Review by jlvargasme — CHANGES_REQUESTED** [review]

File: test/marvelus/function_test.zls (hunk 4 above)
Comment (jchen-cs) [reply test/marvelus/function_test.zls:115]: Looks like most of the 42's are coming from `prove_function`, line 1073, `z3refinement.ml`(not modified by this PR). Perhaps something hasn't been fully implemented there? It seems like all functions get turned into `42` regardless of whether or not they exist in the environment. There's also a TODO there: "check if arguments obey constraints". Does this PR address this TODO comment? 

Edit: This could be that the PR's scope is only function definitions, rather than function applications. If so, can a note be made somewhere to that effect?

File: compiler/verif/z3refinement.ml [hunk 5]
```diff
@@ -1724,35 +1742,30 @@ and vc_gen_typ_exp_desc ctx env typenv t n_source =
```
Comment (jchen-cs) [inline compiler/verif/z3refinement.ml:1742]: Was `n_source` added recently? Can you add a description in the comment?

File: test/marvelus/function_test.ml
Comment (jchen-cs) [inline test/marvelus/function_test.ml:1]: This is a compiled file that should not be included in the PR. You can remove it with `git rm --cached function_test.ml` and consider adding `*.ml` to the `.gitignore` file for this directory.

File: test/marvelus/function_test.zls [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+   The multi_opertation_compute function takes two positive float values, `x` and `y`,
+   and returns a float value `z` such that `z = 2.0 * x + sqrt(x * x - 4.0 * y) / 2.0`.
+   This function should pass the refinement verification.
+    # (=>(and (= v z) (= z (+ (/* 2.0 x) (/ (sqrt (- (/* x x) (/* 4.0 y))) 2.0))) (> y 0.0) (> x 0.0)) (and (>= z 0.0))
+   Constraints:
+   - `x` and `y` must be positive floats.
+   - The return value `z` should satisfy the constraint `z >= 0.0`.
+*)
+let multi_opertation_compute (x : {v:float|v>0.0} ) (y : {v:float|v>0.0} ) : { v:float | v >= 0.0 } =
+  let z = (2.0 *. x) +. (sqrt (x *. x -. 4.0 *. y)) /. 2.0 in
```
Comment (jchen-cs) [inline test/marvelus/function_test.zls:112]: I don't think `sqrt` is implemented yet in the verifier, which is causing `42` to appear in some constraints. This is a separate issue that also needs to be addressed.

File: test/marvelus/function_test.zls (hunk 4 above)
Comment (jchen-cs) [reply test/marvelus/function_test.zls:115]: I'm also noticing `Eseq` causing 42's to appear, which combined with the `42` issue with the functions above, results in a bunch of magic numbers appearing. May not be relevant to include in the current PR but still needs to be addressed.

File: compiler/verif/z3refinement.ml [hunk 7]
```diff
@@ -1721,23 +1742,30 @@ and vc_gen_typ_exp_desc ctx env typenv t =
...
+  | Etypetuple(txp_list) -> 
+    debug(Printf.sprintf "Etypetuple\n"); 
+    List.iter (fun x -> vc_gen_typ_exp_desc ctx env typenv x None) txp_list
+    | Etypevec(texp , si) -> debug(Printf.sprintf "Etypevec\n")
+    | Etypefun(k, t, texp, texp2) -> debug(Printf.sprintf "Etypefun\n")
+    | Etypefunrefinement(k, t, te, te2, e) -> debug(Printf.sprintf "Etypefunrefinement\n")
+    | Erefinement(t, e) -> 
+      debug(Printf.sprintf "Erefinement\n");
+      let expr = vc_gen_expression ctx env e typenv in
+      (* get the variable name from the constriant and replace them by the reference variable *)
```
Comment (jchen-cs) [inline compiler/verif/z3refinement.ml:1760]: I think it might just be a terminology thing, but the code seems to replace the refinement variable with a variable from the function body, rather than a reference variable

File: test/marvelus/function_test.zls
Comment (jchen-cs) [inline test/marvelus/function_test.zls:1]: I noticed that all functions have a `let z=... in z` type of structure. Do other types of function bodies work (e.g. if the function just returns `x +. 1`?

File: main.ml
Comment (jchen-cs) [inline main.ml:1]: This file doesn't appear to be relevant to the PR, was it unintentionally added?

**Review by jchen-cs — CHANGES_REQUESTED** [review]
Good progress, though I have a few questions about specific areas of code (view the comments for details). Looks like we still have to implement VC generation for function application, though that could be out of scope for this specific PR (if it's just handling functions), though this should be made clear in the description.

File: test/marvelus/function_test.zls (hunk 6 above)
Comment (jlvargasme) [reply test/marvelus/function_test.zls:112]: Agreed, we can have a review after this one creating exceptions instead of proceeding with 42. This will help track down the issue and fix bugs.

File: test/marvelus/function_test.zls (hunk 4 above)
Comment (jlvargasme) [reply test/marvelus/function_test.zls:115]: Yeah, if the scope is function definition then we can further investigate this issue in a separate pr. Creating a GitHub tasks for this.

Comment (jlvargasme) [reply test/marvelus/function_test.zls:115]: I commented out the function application on the tests and only looked at function declarations for this review. The only issue I see (which is more like a design choice than an issue), is that the output type of a function is tied to a variable in the verification condition (v = z & x <= v & y <= v), instead of applying the substitution of the variable directly in the boolean constraint (x <= z & y <= z). But I think this works assuming the return value is a single variable (not pairs, with each element typed separately).

`
let add (x : {v:float|v>0.0} ) (y : {v:float|v>0.0} ) : { v:float | x <= v && y <= v } =
  let z = x +. y in
  z
`
`
Proving constraint: (let ((a!1 (=> (and (= v z) (= z (+ x y)) (> y 0.0) (> x 0.0))
               (and (<= x v) (<= y v)))))
  (not a!1))
`
